ezmlm-archive.1
.\" $Id$
.TH ezmlm-archive 1
.SH NAME
ezmlm-archive \- create thread and author index for a mailing list archive
.SH SYNOPSIS
.B ezmlm-archive
[
.B \-cCFTvV
][
.B \-f\fI msg1
]
][
.B \-t\fI msg2
]
.I dir
.SH DESCRIPTION
.B ezmlm-archive
reads the index files from a message archive, and creates a subject index, a
collection of subject files, and a collection of author files. These
files are suitable as an index for WWW access to, and navigation through
a mailing list archive by
.BR ezmlm-cgi(1) .

The index files read are created by
.B ezmlm-idx(1)
on a per-list basis and by
.B ezmlm-send(1)
on a per-message archive for a indexed list.

The output files created are:
.TP
.I dir\fB/archive/threads/yyyymm
The thread index. It contains one line per subject, starting with the
number of the first message with that subject within the set
investigated, ``:'', a 20 character
subject hash, blank, ``[n]'' where ``n'' is the number of messages in the
thread, blank, and the subject.
The file ``yyyymm'' contains
entries for all threads that have messages in the month ``yyyymm''
or that have messages both before and after that month.
The subject hash is a key to the subject files; the message number is
a key to the index file.
The lines are in ascending order by message number when the index is
created
.I de novo
on an existing archive. When the messages are added one-by-one as in normal
archive operation, ``n'' is the number of message in the thread
.I for the particular month
and the order is in reverse of latest message, i.e. the last extended thread
is shown last. The message number accompanying a thread is
always a message within the thread. It is the first in
archives created
on existing lists, and the last message in incrementally created archives.
Use the corresponding subject index file to get a list of all
messages in the thread in ascending order.
.TP
.I dir\fB/archive/subjects/xx/yyyyyyyyyyyyyyyyyy
A subject file. The first line is the subject hash, a space, and the subject.
This is followed by one line per message with this subject, in the format
message number, ``:'', date (yyyymm), ``:'',
author hash, blank, author from line. The lines are
sorted by message number. The author hash is a key to the author files;
the message number is a key to the index file. The file in the example
would be for the subject hash ``xxyyyyyyyyyyyyyyyyyy''.
.TP
.I dir\fB/archive/authors/xx/yyyyyyyyyyyyyyyyyy
An author file. The first line is the author hash, a space, and the author
from line.
This is followed by one line per message with this author, in the format
message number, ``:'', date (yyyymm), ``:'',
subject hash, blank, subject. The lines are
sorted by message number. The subject hash is a key to the subject files;
the message number is a key to the index file. The file in the example
would be for the author hash ``xxyyyyyyyyyyyyyyyyyy''.

.I dir\fB/archnum
keeps track of the last message processed. Normally,
.B ezmlm-archive
will process entries for messages from one above the contents of this file
up to an including the message number in
.IR dir\fB/num .
.SH OPTIONS
.B ezmlm-archive
writes messages in a crash-proof manner when run in normal mode. When overriding
the normal message range with any of the options listed, the normal
.B sync(3)
of the output files is suppressed for efficiency. Should the computer crash
during this time the state of the indices is not defined. Use the
.B \-s
option in the (extremely rare) cases where this would be a problem.
.TP
.B \-c
Create a new index. This  overrides
.I dir\fB/archnum
causing
.B ezmlm-archive
to start with the first message in the archive. Synonym for
.BR \-f\fI0 .
.B NOTE:
.B ezmlm-archive
does not remove files in the index. While it will overwrite/update old files
it will not remove files that are obsolete for other reasons.
.TP
.B \-C
(Default.)
Process entries starting with the message after the message listed in
.IR dir\fB/archnum .
.TP
.B \-f\fI msg1
Process messages from the archive section (set of 100 messages)
containing message
.IR msg1 .
This is useful if you have removed part of the archive, as it will shorten
processing time and decrease memory use.
.B NOTE:
.B ezmlm-archive
does not remove files in the index. While it will overwrite/update old files
it will not remove files that are obsolete for other reasons. The number of
messages per thread will be incorrect when using of the
.B \-f
and
.B \-t
switches leads to partial re-indexing of already indexed messages.
.TP
.B \-F
(Default.)
Do not change the starting message from the default
(see
.BR \-C ).
.TP
.B \-s
Always sync files.
.TP
.B \-S
(Default.)
Sync files, except when on of the message range modifying options is
used.
.TP
.B \-t\fI msg2
Process messages to message
.I msg2
instead of the last message in the archive. Again, files written are
corrected, but other files are not explicitly removed.
.TP
.B \-T
(Default.)
Process entries for messages up to the last message in the archive.
.TP
.B \-v
Display
.B ezmlm-archive
version info.
.TP
.B \-V
Display
.B ezmlm-archive
version info.
.SH "MEMORY USAGE"
.B ezmlm-archive
stores its linked lists in memory. On at 32-bit architecture, it uses
12 bytes per message, 28 bytes per thread (plus one copy of the subject),
and 20 bytes per author (plus one copy of the author from line).

In normal list use, it processes only at most a few messages at a time,
but for initial processing of a large archive, considerable amounts of
memory may be used. Assuming
40 bytes for subject/from line, 5 messages per thread, 100,000 messages,
and 1000 authors, this is 2.5 MB. For 1,000,000 messages this is about 20 MB.

Thus, for large archives, it may be useful to use the
.I \-t
switch to process the archive in multiple subsets, starting with e.g. the first
100,000, then the next, and so on.
.SH "SEE ALSO"
ezmlm-cgi(1),
ezmlm-idx(1),
ezmlm-send(1),
ezmlm(5)



CHANGES
BUGS:	-ezmlm-idx does not implement a full rfc822 parser. Comments can be
	 inserted to defeat the MIME parsing mechanism, such as:
	 Content-Type: (comment) test/plain
	 This type of comment, while legal, does not occur in normal messages.
	 Guiding info past ezmlm-idx MIME parsing is easier done by enclosing
	 a message within a message. The mechanism isn't intended as a
	 "secure" way of preventing e.g. "application/ms-tnef" from being
	 disseminated.
	 If you think this is a problem, E-mail ezmlm@list.cr.yp.to.
	 Mimeremove/reject work only at the outer MIME layer.
	-ezmlm-request requires addresses to be user@host. While rfc822
	 compliant, addresses like ``"the man" @ host'' are not accepted.
	-ezmlm-cgi doesn't deal optimally with holes in the archive.
	-ezmlm-cgi seems not to work with Chinese. The problem has been to
	 find someone willing/able to participate in debugging.

Some ezmlmrc are not updated and may not work right.  If you fix them,
please send them to ezmlm@list.cr.yp.to for inclusion in the next
release.

ezmlm-idx-7.1.0, ????-??-??
===========================

- Added support for SQLite3 subscriber databases.
  Thanks to Mike Tedder.

- Added support for wildcard addresses in all address databases.

- Added support for qmqpservers to all ezmlm-idx programs.

- Added support for decoding sender addresses mangled with a BATV "btv1" tag.

- Modified the (un)subscribe procedure to not require confirmation
  before telling the sender they were already on (or off) the list.

- Renamed to "setup" target to the more standard "install".

- Modified ezmlm-split to exit silently if there is not "split" file,
  and added it back into the "manager" file to properly handle subscribe
  and unsubscribe requests destined for sublists.

- Made the ezmlm-reject -h (obey headerreject) option the default.

- Fixed ezmlm-send to obey the "addtrailer" setting instead of looking
  for a "text/trailer" file.

- Added support for custom subject lines to ezmlm-get (for digests).

- Fixed handling of adding the trailer on old lists.

- Fixed adding the trailer to posts encoded with base64.

ezmlm-idx-7.0.2, 2009-08-09
===========================

- Fixed selection of [un]subscribe subject line for moderated
  subscriptions.

- Clarified ezmlm-reject man page to note that Precedence: bulk is also
  rejected.

- Added a proper charset for ch_GB (Chinese GuoBiao), and added an
  explicit charset for the other languages that were missing them.

- Switched all ISO-8859-1 charsets to the more modern ISO-8859-15.

- Many small text updates and tweaks.

ezmlm-idx-7.0.1, 2009-07-09
===========================

- Added support for decoding sender addresses mangled by BATV prvs.

- Added missing ezmlm-checksub to the installed programs.

- Improved the TXT_BY message used in digests to work better with
  non-English languages.

- Fixed out-of-memory resulting in a permanent error.

- Fixed a buffer underflow bug in concatHDR.

- Fixed a bug in ezmlm-issubn.c that was causing it to scan the main
  list when it shouldn't.

- Fixed missing defaults in ezmlm-tstdig when digsize, digcount, or
  digtime were not created.

- Clarified the steps necessary when upgrading in UPGRADE.

ezmlm-idx-7.0.0, 2008-06-16
===========================

This version has three major architectural changes:

1. There is now support for internationalized messages.  All error
messages, all subject lines, and a few other messages are now configured
in a file named "text/messages" that is read at run time.  Messages in
this file will undergo header-style substitution before they are output.

Unlike other text files, all 3 of the files (DIR/text/messages,
/etc/ezmlm/LANG/text/messages, and /etc/ezmlm/default/text/messages) are
read, and only the first match (in the above order) is used for any
given message.  This allows for creation of partial files to override
just select messages.  If no match is found, the internal English text
is used.

2. The use of the "flags" file has been deprecated completely by
individual flag files accessed by the appropriate programs.  This, along
with corresponding changes to the ezmlmrc template, allows for lists to
be reconfigured without invoking ezmlm-make.  To upgrade a list to the
new setup, simply run "ezmlm-make -+ DIR".  Note that this will delete
the "config" and "flags" files, as they have been superceded by other
files.

3. Creation of subscriber tables has been moved out of ezmlm-mktab-* and
into ezmlm-make by adding some additional hooks to the subdb plugins.  A
ezmlm-rmtab program is also added to remove subscriber tables, also with
hooks in the subdb plugins.  The ezmlm-mktab-* programs have been
obsoleted by these changes and have been removed.

Smaller changes:

- Fixed a bug in ezmlm-request that caused a segfault when attempting to
  execute the "which" command.

- Added new program ezmlm-checksub to replace the
	ezmlm-issubn ... || { echo error; exit 100; }
  lines in the ezmlmrc files.  ezmlm-issubn is now deprecated for use in
  .qmail files, but still exists for backwards compatibility.

- Added ezmlm-weed to the confirmer control files.

- Added <#a#> substitution for the local part of the accept address.

- ezmlm-warn will now process bounces for both the main and digest lists
  if neither the -d nor the -D options are given.

- Added several control files for programs:
	dir/digcount		ezmlm-tstdig
	dir/digestcode		ezmlm-get
	dir/digformat		ezmlm-get
	dir/digsize		ezmlm-tstdig
	dir/digtime		ezmlm-tstdig
	dir/modcanedit		ezmlm-manage
	dir/modcanlist		ezmlm-manage
	dir/modgetonly		ezmlm-get
	dir/modpostonly		ezmlm-store
	dir/nosubconfirm	ezmlm-manage
	dir/nounsubconfirm	ezmlm-manage
	dir/nowarn		ezmlm-warn
	dir/subgetonly		ezmlm-get
	dir/noreturnposts	ezmlm-clean

- Removed the ezmlm-make -4 option in favor of the above files.

- Fixed handling of digest bouncer in ezmlm-dispatch.

- Eliminated the use of vfork for portability.


ezmlm-archive.c
/*$Id$*/
#include <sys/types.h>
#include <sys/stat.h>
#include <stdio.h>
#include <unistd.h>
#include "alloc.h"
#include "error.h"
#include "stralloc.h"
#include "str.h"
#include "sig.h"
#include "getconf.h"
#include "strerr.h"
#include "getln.h"
#include "substdio.h"
#include "readwrite.h"
#include "fmt.h"
#include "sgetopt.h"
#include "idxthread.h"
#include "makehash.h"
#include "lock.h"
#include "open.h"
#include "scan.h"
#include "idx.h"
#include "errtxt.h"
#include "auto_version.h"

const char FATAL[] = "ezmlm-archive: fatal: ";
#define WARNING "ezmlm-archive: warning: inconsistent index: "

substdio ssin;
char inbuf[1024];
substdio ssout;
char outbuf[1024];
substdio ssnum;
char numbuf[16];

stralloc line = {0};
stralloc num = {0};
stralloc fn = {0};
stralloc fnn = {0};

char strnum[FMT_ULONG];
int flagerror = 0;
int flagsync = 1;	/* sync() by default, not for -c or -f or -t */
char *dir;

struct ca {
  char *s;		/* start */
  unsigned int l;	/* length */
} ca;

void die_usage() {
  strerr_die1x(100,
    "ezmlm-archive: usage: "
	"ezmlm-archive [-cCFsSTvV] [-f min_msg] [-t max_msg] dir");
}

void die_nomem() { strerr_die2x(111,FATAL,ERR_NOMEM); }

void close_proper(ss,s,sn)
/* flush,sync,close,move sn->s) */
substdio *ss;
char *s, *sn;
{
   if (substdio_flush(ss) == -1)
     strerr_die6sys(111,FATAL,ERR_FLUSH,dir,"/",s,": ");
  if (flagsync)
    if (fsync(ss->fd) == -1)
       strerr_die6sys(111,FATAL,ERR_SYNC,dir,"/",s,": ");
  if (close(ss->fd) == -1)
     strerr_die6sys(111,FATAL,ERR_CLOSE,dir,"/",s,": ");
  if (rename(sn,s) == -1)
     strerr_die6sys(111,FATAL,ERR_MOVE,dir,"/",sn,": ");
}

void write_threads(msgtable,subtable,authtable,datetable,from,to)
/* Add the current threading data to the thread database without dups */
/* Writes the subject index first, then processes the individual files */
msgentry *msgtable; subentry *subtable; authentry *authtable;
dateentry *datetable;
unsigned long from,to;
{
  msgentry *pmsgt;
  subentry *psubt,*psubtm, *psubtlast;
  subentry *presubt = (subentry *)0;
  authentry *pautht;
  dateentry *pdatet;
  const char *cp;
  const char *cp1;
  unsigned long msg;
  unsigned long ulmsginthread;
  unsigned long subnum;
  unsigned long authnum;
  unsigned long msgnum;
  unsigned int pos;
  unsigned int startdate,nextdate;
  unsigned int startmsg,nextmsg;
  int fd = -1;
  int fdn = -1;
  int match;
  int ffound;
  int lineno;
  int res;

  psubtm = subtable;		/* now for new threads */
  pdatet = datetable;
  nextmsg = 0L;
  nextdate = pdatet->date;
  while (psubtm->sub) {		/* these are in msgnum order */
    if (!presubt)		/* for rewind */
      if (psubtm->lastmsg >= nextmsg)
	presubt = psubtm;	/* this thread extends beyond current month */
    if (psubtm->firstmsg >= nextmsg) {	/* done with this month */
      if (fdn != -1) close_proper(&ssout,fn.s,fnn.s);
      psubtlast = psubtm;		/* last thread done */
      if (presubt)			/* need to rewind? */
	psubtm = presubt;		/* do it */
      psubt = psubtm;			/* tmp pointer to reset done flag */
      presubt = (subentry *)0;		/* reset rewind pointer */
      pdatet++;				/* next month */
      startdate = nextdate;		/* startdate */
      nextdate = pdatet->date;		/* end date */
      startmsg = nextmsg;		/* first message in month */
      nextmsg = pdatet->msg;		/* first message in next month */
      if (!stralloc_copys(&fn,"archive/threads/")) die_nomem();
      if (!stralloc_catb(&fn,strnum,fmt_uint(strnum,startdate))) die_nomem();
      if (!stralloc_copy(&fnn,&fn)) die_nomem();
      if (!stralloc_0(&fn)) die_nomem();
      if (!stralloc_cats(&fnn,"n")) die_nomem();
      if (!stralloc_0(&fnn)) die_nomem();
      if ((fdn = open_trunc(fnn.s)) == -1)
	strerr_die6sys(111,FATAL,ERR_CREATE,dir,"/",fnn.s,": ");
      substdio_fdbuf(&ssout,write,fdn,outbuf,sizeof(outbuf));
      if ((fd = open_read(fn.s)) == -1) {
      if (errno != error_noent)
             strerr_die6sys(111,FATAL,ERR_OPEN,dir,"/",fn.s,": ");
      } else {
	substdio_fdbuf(&ssin,read,fd,inbuf,sizeof(inbuf));
      for (;;) {
      if (getln(&ssin,&line,&match,'\n') == -1)
             strerr_die6sys(111,FATAL,ERR_READ,dir,"/",fn.s,": ");
      if (!match) break;
      pos = scan_ulong(line.s,&msgnum);
      pos++;			/* skip ':' */
      if (msgnum >= from)
	continue;		/* ignore entries from threading range */
      if (line.len < pos + HASHLEN) {
	flagerror = -1;		/* and bad ones */
	continue;
      }
      psubt = subtable;
      cp = line.s + pos;
      ffound = 0;		/* search among already known subjects */
      for (;;) {
	res = str_diffn(psubt->sub,cp,HASHLEN);
	if (res < 0) {
	  if (psubt->higher)
	    psubt = psubt->higher;
	 else
	   break;
	} else if (res > 0) {
	  if (psubt->lower)
	    psubt = psubt->lower;
	  else
	    break;
	} else {
	  ffound = 1;
	  break;
	}
      }
      if (!ffound) {
	if (substdio_put(&ssout,line.s,line.len) == -1)
	  strerr_die6sys(111,FATAL,ERR_WRITE,dir,"/",fnn.s,": ");
      } else {			/* new # of msg in thread */
	cp += HASHLEN;		/* HASHLEN [#] Subject always \n at end */
	if (*(cp++) == ' ' && *(cp++) == '[') {
	  cp += scan_ulong(cp,&ulmsginthread);
	  if (*cp == ']') {
	    psubt->msginthread += (unsigned char) (ulmsginthread & 0xff);
	  }
	} else
	  flagerror = -5;
      }
    }
    close(fd);
  }
  continue;
  }

    if (psubtm->firstmsg < nextmsg && psubtm->lastmsg >= startmsg) {
    if (!stralloc_copyb(&line,strnum,fmt_ulong(strnum,psubtm->lastmsg)))
		die_nomem();
    if (!stralloc_cats(&line,":")) die_nomem();
    if (!stralloc_catb(&line,psubtm->sub,HASHLEN)) die_nomem();
    if (!stralloc_cats(&line," [")) die_nomem();
    if (!stralloc_catb(&line,strnum,
	fmt_ulong(strnum,(unsigned long) psubtm->msginthread)))
		die_nomem();
    if (!stralloc_cats(&line,"]")) die_nomem();
    if (!stralloc_catb(&line,psubtm->sub + HASHLEN,psubtm->sublen - HASHLEN))
			 die_nomem();	/* has \n */
    if (substdio_put(&ssout,line.s,line.len) == -1)
	strerr_die6sys(111,FATAL,ERR_WRITE,dir,"/",fnn.s,": ");
    }
  psubtm++;
  }
  if (fdn != -1)
    close_proper(&ssout,fn.s,fnn.s);

  psubt = subtable;
  while (psubt->sub) {		/* now the threads */
    if (!stralloc_copys(&fn,"archive/subjects/")) die_nomem();
    if (!stralloc_catb(&fn,psubt->sub,2)) die_nomem();
    if (!stralloc_0(&fn)) die_nomem();
    if (mkdir(fn.s,0755) == -1)
    if (errno != error_exist)
      strerr_die6sys(111,FATAL,ERR_CREATE,dir,"/",fn.s,": ");
    fn.s[fn.len - 1] = '/';
    if (!stralloc_catb(&fn,psubt->sub+2,HASHLEN-2)) die_nomem();
    if (!stralloc_copy(&fnn,&fn)) die_nomem();
    if (!stralloc_cats(&fnn,"n")) die_nomem();
    if (!stralloc_0(&fn)) die_nomem();
    if (!stralloc_0(&fnn)) die_nomem();
    if ((fdn = open_trunc(fnn.s)) == -1)
      strerr_die4sys(111,FATAL,ERR_CREATE,fnn.s,": ");
    substdio_fdbuf(&ssout,write,fdn,outbuf,sizeof(outbuf));
    if ((fd = open_read(fn.s)) == -1) {
      if (errno != error_noent)
	  strerr_die4sys(111,FATAL,ERR_OPEN,fn.s,": ");
      if (substdio_puts(&ssout,psubt->sub) == -1)	/* write subject */
	     strerr_die6sys(111,FATAL,ERR_WRITE,dir,"/",fnn.s,": ");
    } else {					/* copy data */
	substdio_fdbuf(&ssin,read,fd,inbuf,sizeof(inbuf));
	lineno = 0;
	for (;;) {
	  if (getln(&ssin,&line,&match,'\n') == -1)
             strerr_die6sys(111,FATAL,ERR_READ,dir,"/",fn.s,": ");
          if (!match) break;
	  if (!lineno) {			/* write subject */
	    if (line.len < HASHLEN + 1 || line.s[HASHLEN] != ' ')
		flagerror = -3;
	    if (substdio_put(&ssout,line.s,line.len) == -1)
	       strerr_die6sys(111,FATAL,ERR_WRITE,dir,"/",fnn.s,": ");
	    lineno = 1;
	    continue;
	  }
	  (void) scan_ulong(line.s,&msgnum);
	  if (msgnum >= from) break;
	  if (substdio_put(&ssout,line.s,line.len) == -1)
	     strerr_die6sys(111,FATAL,ERR_WRITE,dir,"/",fnn.s,": ");
	}
	(void) close(fd);	/* close old index */
      }

    subnum = (unsigned long) (psubt - subtable + 1);	/* idx of this subj */
    pmsgt = msgtable + psubt->firstmsg - from;	/* first message entry */
    for (msg = psubt->firstmsg; msg <= psubt->lastmsg; msg++) {
      if (pmsgt->subnum == subnum) {
        if (!stralloc_copyb(&line,strnum,fmt_ulong(strnum,msg))) die_nomem();
        if (!stralloc_cats(&line,":")) die_nomem();
	if (!stralloc_catb(&line,strnum,fmt_uint(strnum,pmsgt->date)))
		die_nomem();
	if (!stralloc_cats(&line,":")) die_nomem();
        if (pmsgt->authnum) {
	  pautht = authtable + pmsgt->authnum - 1;
	  cp = pautht->auth;
	  cp1 = cp + str_chr(cp,' ');
	  if (cp + HASHLEN != cp1)
	    strerr_die1x(100,ERR_BAD_INDEX);
	  if (!stralloc_cats(&line,cp))
		die_nomem();				/* hash */
	} else
          if (!stralloc_cats(&line,"\n")) die_nomem();
	if (substdio_put(&ssout,line.s,line.len) == -1)
	  strerr_die6sys(111,FATAL,ERR_WRITE,dir,"/",fnn.s,": ");
      }
      pmsgt++;
    }
    close_proper(&ssout,fn.s,fnn.s);
    psubt++;
  }

					/* (no master author index) */
  pautht = authtable;
  while (pautht->auth) {		/* now the authors */
    if (!stralloc_copys(&fn,"archive/authors/")) die_nomem();
    if (!stralloc_catb(&fn,pautht->auth,2)) die_nomem();
    if (!stralloc_0(&fn)) die_nomem();
    if (mkdir(fn.s,0755) == -1)
    if (errno != error_exist)
      strerr_die6sys(111,FATAL,ERR_CREATE,dir,"/",fn.s,": ");
    fn.s[fn.len - 1] = '/';
    if (!stralloc_catb(&fn,pautht->auth+2,HASHLEN-2)) die_nomem();
    if (!stralloc_copy(&fnn,&fn)) die_nomem();
    if (!stralloc_cats(&fnn,"n")) die_nomem();
    if (!stralloc_0(&fn)) die_nomem();
    if (!stralloc_0(&fnn)) die_nomem();
    if ((fdn = open_trunc(fnn.s)) == -1)
      strerr_die4sys(111,FATAL,ERR_CREATE,fnn.s,": ");
    substdio_fdbuf(&ssout,write,fdn,outbuf,sizeof(outbuf));
      if ((fd = open_read(fn.s)) == -1) {
	if (errno != error_noent)
	  strerr_die4sys(111,FATAL,ERR_OPEN,fn.s,": ");
        else {			/* didn't exist before: write author */
          if (substdio_put(&ssout,pautht->auth,pautht->authlen) == -1)
	     strerr_die6sys(111,FATAL,ERR_WRITE,dir,"/",fnn.s,": ");
	}
      } else {					/* copy data */
	substdio_fdbuf(&ssin,read,fd,inbuf,sizeof(inbuf));
	lineno = 0;
	for (;;) {
	  if (getln(&ssin,&line,&match,'\n') == -1)
             strerr_die6sys(111,FATAL,ERR_READ,dir,"/",fn.s,": ");
          if (!match) break;
	  if (!lineno) {			/* write author */
	    if (line.len < HASHLEN + 1 || line.s[HASHLEN] != ' ')
		flagerror = - 4;
	    if (substdio_put(&ssout,line.s,line.len) == -1)
	       strerr_die6sys(111,FATAL,ERR_WRITE,dir,"/",fnn.s,": ");
	    lineno = 1;
	    continue;
	  }
	  (void) scan_ulong(line.s,&msgnum);
	  if (msgnum >= from) break;
	  if (substdio_put(&ssout,line.s,line.len) == -1)
	     strerr_die6sys(111,FATAL,ERR_WRITE,dir,"/",fnn.s,": ");
	}
	(void) close(fd);			/* close old index */
      }

    authnum = (unsigned long) (pautht - authtable + 1);	/* idx of this auth */
    pmsgt = msgtable + pautht->firstmsg - from;	/* first message entry */
    for (msg = pautht->firstmsg; msg <= to; msg++) {
      if (pmsgt->authnum == authnum) {
        if (!stralloc_copyb(&line,strnum,fmt_ulong(strnum,msg))) die_nomem();
        if (!stralloc_cats(&line,":")) die_nomem();
	if (!stralloc_catb(&line,strnum,fmt_uint(strnum,pmsgt->date)))
		die_nomem();
	if (!stralloc_cats(&line,":")) die_nomem();
        if (pmsgt->subnum) {
	  psubt = subtable + pmsgt->subnum - 1;
          if (!stralloc_catb(&line,psubt->sub,psubt->sublen))
		die_nomem();
	}
	if (substdio_put(&ssout,line.s,line.len) == -1)
	  strerr_die6sys(111,FATAL,ERR_WRITE,dir,"/",fnn.s,": ");
      }
      pmsgt++;
    }
    close_proper(&ssout,fn.s,fnn.s);
    pautht++;
  }
}

int main(argc,argv)
int argc;
char **argv;
{
  unsigned long archnum = 0L;
  unsigned long to = 0L;
  unsigned long max;
  int fd;
  int fdlock;
  int flagcreate = 0;
  int flagsyncall = 0;
  int opt;
  msgentry *msgtable;
  subentry *subtable;
  authentry *authtable;
  dateentry *datetable;

  (void) umask(022);
  sig_pipeignore();

  while ((opt = getopt(argc,argv,"cCf:FsSt:TvV")) != opteof)
    switch (opt) {
      case 'c':	flagcreate = 1;
		flagsync = 0;
		break;			/* start at beginning of archive */
      case 'C': flagcreate = 0;
		break;	/* Do only archnum+1 => num */
      case 'f': if (optarg) {
		  (void) scan_ulong(optarg,&archnum);
		  archnum = (archnum / 100) * 100;
	        }
		flagsync = 0;
		break;
      case 'F': archnum = 0; break;
      case 's': flagsyncall = 1; break;
      case 'S': flagsyncall = 0; break;
      case 't': if (optarg) {
		  (void) scan_ulong(optarg,&to);
		}
		flagsync = 0;
		break;
      case 'T': to = 0; break;
      case 'v':
      case 'V': strerr_die2x(0,"ezmlm-archive version: ",auto_version);
      default:
        die_usage();
    }

  if (flagsyncall) flagsync = 1;	/* overrides */
  dir = argv[optind++];
  if (!dir) die_usage();
  if (chdir(dir) == -1)
    strerr_die4sys(111,FATAL,ERR_SWITCH,dir,": ");

  if (mkdir("archive/threads",0755) == -1)
    if (errno != error_exist)
      strerr_die4sys(111,FATAL,ERR_CREATE,dir,"/archive/threads: ");
  if (mkdir("archive/subjects",0755) == -1)
    if (errno != error_exist)
      strerr_die4sys(111,FATAL,ERR_CREATE,dir,"/archive/subjects: ");
  if (mkdir("archive/authors",0755) == -1)
    if (errno != error_exist)
      strerr_die4sys(111,FATAL,ERR_CREATE,dir,"/archive/authors: ");

	/* Lock list to assure that no ezmlm-send is working on it */
	/* and that the "num" message is final */
  fdlock = open_append("lock");
  if (fdlock == -1)
    strerr_die2sys(111,FATAL,ERR_OPEN_LOCK);
  if (lock_ex(fdlock) == -1) {
    (void) close(fdlock);
    strerr_die2sys(111,FATAL,ERR_OBTAIN_LOCK);
  }
					/* get num */
  if (!getconf_line(&num,"num",0,dir))
    strerr_die1x(100,ERR_EMPTY_LIST);
  (void) close(fdlock);

  if (!stralloc_0(&num)) die_nomem();	/* parse num */
  (void) scan_ulong(num.s,&max);
  if (!to || to > max) to = max;

  fdlock = open_append("archive/lock");	/* lock index */
  if (fdlock == -1)
    strerr_die4sys(111,FATAL,ERR_OPEN,dir,"/archive/lock: ");
  if (lock_ex(fdlock) == -1) {
    (void) close(fdlock);
    strerr_die4sys(111,FATAL,ERR_OBTAIN,dir,"/archive/lock: ");
  }
  if (!flagcreate && !archnum) {	/* adjust archnum (from) / to */
    if (getconf_line(&num,"archnum",0,dir)) {
      if (!stralloc_0(&num)) die_nomem();
      (void) scan_ulong(num.s,&archnum);
      archnum++;
    }
  }

  if (archnum > to)
    _exit(0);				/* nothing to do */

					/* do the subject threading */
  idx_mkthreads(&msgtable,&subtable,&authtable,&datetable,archnum,to,max,0);
					/* update the index */
  write_threads(msgtable,subtable,authtable,datetable,archnum,to);
					/* update archnum */
  if ((fd = open_trunc("archnumn")) == -1)
    strerr_die4sys(111,FATAL,ERR_CREATE,dir,"/archnumn: ");
  substdio_fdbuf(&ssnum,write,fd,numbuf,sizeof(numbuf));
  if (substdio_put(&ssnum,strnum,fmt_ulong(strnum,to)) == -1)
     strerr_die6sys(111,FATAL,ERR_WRITE,dir,"/",fnn.s,": ");
  if (substdio_puts(&ssnum,"\n") == -1)
     strerr_die6sys(111,FATAL,ERR_WRITE,dir,"/",fnn.s,": ");
  close_proper(&ssnum,"archnum","archnumn");
  switch (flagerror) {
    case 0:
       _exit(0);				/* go bye-bye */
    case -1:
       strerr_die2x(99,WARNING,"threads entry with illegal format");
    case -2:
       strerr_die2x(99,WARNING,"thread in index, but threadfile missing");
    case -3:
       strerr_die2x(99,WARNING,"a subject file lacks subject");
    case -4:
       strerr_die2x(99,WARNING,"an author file lacks author/hash");
    case -5:
       strerr_die2x(99,WARNING,"threads entry lacks message count");
    default:
       strerr_die2x(99,WARNING,"something happened that isn't quite right");
  }
}

